packages/core/src/hydration/error_handling.ts

```diff
@@ -6,76 +6,422 @@
+  const footer = getHydrationErrorFooter();
+
+  throw new RuntimeError(RuntimeErrorCode.HYDRATION_MISSING_NODE, header + expected + footer);
 }
 
+/**
+ * Builds a hydration error message when a node is not found at a path location
+ *
+ * @param host the Host Node
+ * @param path the path to the node
+ */
 export function nodeNotFoundAtPathError(host: Node, path: string): Error {
-  // TODO: improve error message and use RuntimeError instead.
-  return new Error('During hydration Angular was unable to locate a node');
+  const header = `During hydration Angular was unable to locate a node ` +
+      `using the "${path}" path, starting from the ${describeRNode(host)} node.\n\n`;
+  const footer = getHydrationErrorFooter();
+
+  throw new RuntimeError(RuntimeErrorCode.HYDRATION_MISSING_NODE, header + footer);
 }
 
-export function unsupportedProjectionOfDomNodes(): Error {
-  // TODO: improve error message and use RuntimeError instead.
-  return new Error(
-      'During serialization, Angular detected DOM nodes ' +
+
+/**
+ * Builds the hydration error message in the case that dom nodes are created outside of
+ * the Angular context and are being used as projected nodes
+ *
+ * @param lView the LView
+ * @param tNode the TNode
+ * @returns an error
+ */
+export function unsupportedProjectionOfDomNodes(rNode: RNode): Error {
+  const header = 'During serialization, Angular detected DOM nodes ' +
       'that were created outside of Angular context and provided as projectable nodes ' +
       '(likely via `ViewContainerRef.createComponent` or `createComponent` APIs). ' +
       'Hydration is not supported for such cases, consider refactoring the code to avoid ' +
-      'this pattern or using `ngSkipHydration` on the host element of the component.');
+      'this pattern or using `ngSkipHydration` on the host element of the component.\n\n';
+  const actual = `${describeDomFromNode(rNode)}\n\n`;
+  const message = header + actual + getHydrationAttributeNote();
+  return new RuntimeError(RuntimeErrorCode.UNSUPPORTED_PROJECTION_DOM_NODES, message);
 }
 
-export function invalidSkipHydrationHost() {
-  // TODO: improve error message and use RuntimeError instead.
-  return new Error(
-      'The `ngSkipHydration` flag is applied on a node ' +
+/**
+ * Builds the hydration error message in the case that ngSkipHydration was used on a
+ * node that is not a component host element or host binding
+ *
+ * @param rNode the HTML Element
+ * @returns an error
+ */
+export function invalidSkipHydrationHost(rNode: RNode): Error {
+  const header = 'The `ngSkipHydration` flag is applied on a node ' +
       'that doesn\'t act as a component host. Hydration can be ' +
-      'skipped only on per-component basis.');
+      'skipped only on per-component basis.\n\n';
+  const actual = `${describeDomFromNode(rNode)}\n\n`;
+  const footer = 'Please move the `ngSkipHydration` attribute to the component host element ' +
+      'or host binding.';
```

Comment:
I think we can omit the 'or host binding' part, since it's just a method of setting the host attribute.
```suggestion
  const footer = 'Please move the `ngSkipHydration` attribute to the component host element.';
```